specs/impact-model-plugin.md

```diff
@@ -0,0 +1,335 @@
+
+Currently if you want to consume an IMM you are required to craft custom interfaces, since every software measurement use case is bespoke. Swapping one IMM for another requires code changes and comparing IMMs or validating their accuracy/precision is challenging. 
+
+If every IMM **exposed the same interface**, then those models can easily be plugged in different applications, swapped in and out, upgraded, compared.
+
+**Our thesis is simple, if we want there to be a large, vibrant, ecosystem of people and tooling around Impact measurement then we need a standard, common, interface to all Impact Measurement Models.**
+
+The goal of the Impact Measurement Plugin (IMP) Spec is to define a common standardised interface to every IMM, current and future. 
+
+## Impact Model Interface Specification 
+
+The Impact Model Interface Specification project is here to standardise an interface to existing and future impact models (IMOs). The ecosystem of solutions we need for Impact measurement will sit on a solid bedrock of Impact Model Interfaces (IMIs) that have the same interface, to enable reusability, and comparability. 
+
+Any class library that exposes the IMI interface adheres to this specification.
+
+The interface describes the bidirectional communication between an IMO and what's calling the IMO. It describes the function signature, the data types and the exceptions used to pass back information to the caller.
+
+This version of the specification does not pick any specific language, but it does assume the language has support for exception handling.
+
+## Data Types
+
+### StaticParams
+
+This is a simple dictionary used to set up an IMM. Each IMM is different and can have a wide variety of parameters. The specification cannot define what these parameters will be, so we choose to have a simple data type of a dictionary which accepts all types and quantity of params.
+
+### Telemetry
+
+This is a unit of data to describe some inputs to an IMM. Since every IMM is different we can’t prescribe too much in the specification, however the only two fields that would be mandatory for each unit of telemetry is the date/time when the telemetry was gathered and the time-range for which the telemetry is for. For example, you might have some telemetry for CPU utilisation, but we also need to know when this telemetry was gathered and for what period of time does the telemetry measure.
+
+Telemetry can be aggregate or granular, for example it might contain an average of cpu utilization across a range of machines.
+
+`{ “date-time”: xxxx, “duration”: xxx, “cpu-util”: 0.5 }`
+
+Or it might contain more granular data for every machine like so:
+
+`{ “date-time”: xxxx, “duration”: xxx, machines: { “gsf-001”: “cpu-util”: 0.5, “gsf-002”:  “cpu-util”: 0.62 } }`
+
+Or it might contain even more granular data to the process level:

```

Comment:
We probably need extend the example with machineType, like a Cloud VM type like D2, C2 machines, or On-premises with actual machine spec.

Reply:
Agreed, I think as we start developing Telemetry is going to have much more details, suspect it's going to have it's own spec. I think this will form a summary doc and with that in mind I'd suggest keeping this initial example simple.